### PR TITLE
Parser: parse var as Call evan if var is declared in call args

### LINKED ISSUE #4106: Improve error message for "undefined local variable or method" when variable is previosuly defined because of "var : Type"
```crystal
module Mod
  class Klass
    getter name : String

    def initialize(@name)
    end
  end

  puts name
end
```

raises:

```
Error in test.cr:9: read before assignment to local variable 'name'

  puts name
       ^~~~
```

Crystal 0.21.0, mac os x yosemite

### PR DESCRIPTION
**Pull request by MakeNowJust**
Fixed #4106

### CODE REVIEW

**Comment by MakeNowJust** [comment]
This change is useful in following case for example:

```crystal
class Foo
  class_property foo : String = "hello"

  p foo
end
```

Currently this file causes compile error: read before assignment to local variable 'foo'. But, this behavior looks like a bug.

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -3687,10 +3688,12 @@ module Crystal
       node
     end
 
-    def preserve_stop_on_do(new_value = false)
+    def preserve_stop_on_do(new_value = false, call_args? = true)
```
Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr]: Could you use `is_call_args` instead? `call_args?` as a variable name should be a parse error (at least it's not allowed in Ruby, and in Crystal it looks weird). If we don't change it now we'll have to change it later when we fix the parser for this, but I don't want to encourage this names.

File: src/compiler/crystal/syntax/parser.cr [hunk 2]
```diff
@@ -3687,10 +3688,12 @@ module Crystal
...
       old_stop_on_do = @stop_on_do
       @stop_on_do = new_value
+      @call_args_nest += 1 if call_args?
```
Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr]: Maybe call_args_nest should just be incremented and decremented in the methods `parse_call_args` and `parse_call_args_space_consumed`? I don't think this is the right place to do it.

**Comment by asterite** [comment]
Thank you for this!

I think this change is OK. It's a breaking change, because if you have code similar to the spec you added:

```crystal
macro debug_init(x)
  puts "initializing {{x.var}} to {{x.value}}"
  {{x}}
end

debug_init a : Int32 = 0
a
```

it will stop working (now it works). But I don't think that's a good usage for a macro, and in general we use `var : Type` to eventually declare the type of instance/class vars, never local vars.

I made a few comments, after that I'll review it again and probably merge this.

**Comment by MakeNowJust** [comment]
@asterite Thanks for reviewing. I fixed to incr/decr `@call_args_nest` inside of `parse_call_args` and `parse_call_args_space_consumed`.

**Comment by asterite** [comment]
@MakeNowJust Thank you! 💚

**Pull request merged**